catalog/resource_grant.go

```diff
@@ -0,0 +1,187 @@
+	"sort"
+	"strings"
+
+	"github.com/databricks/databricks-sdk-go/apierr"
+	"github.com/databricks/terraform-provider-databricks/catalog/permissions"
+	"github.com/databricks/terraform-provider-databricks/catalog/util"
+	"github.com/databricks/terraform-provider-databricks/common"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+// diffPermissionsForPrincipal returns UnityCatalogPermissionsDiff of this permissions list with `diff` privileges removed
+func diffPermissionsForPrincipal(principal string, pl permissions.UnityCatalogPermissionsList, existing permissions.UnityCatalogPermissionsList) (diff permissions.UnityCatalogPermissionsDiff) {
+	// diffs change sets for principal
+	configured := map[string]*schema.Set{}
+	for _, v := range pl.Assignments {
+		if v.Principal == principal {
+			configured[v.Principal] = util.SliceToSet(v.Privileges)
+		}
+	}
+	// existing permissions that needs removal for principal
+	remote := map[string]*schema.Set{}
+	for _, v := range existing.Assignments {
+		if v.Principal == principal {
+			remote[v.Principal] = util.SliceToSet(v.Privileges)
+		}
+	}
+	// STEP 1: detect overlaps
+	for principal, confPrivs := range configured {
+		remotePrivs, ok := remote[principal]
+		if !ok {
+			remotePrivs = util.SliceToSet([]string{})
+		}
+		add := util.SetToSlice(confPrivs.Difference(remotePrivs))
+		remove := util.SetToSlice(remotePrivs.Difference(confPrivs))
+		if len(add) == 0 && len(remove) == 0 {
+			continue
+		}
+		diff.Changes = append(diff.Changes, permissions.UnityCatalogPermissionsChange{
+			Principal: principal,
+			Add:       add,
+			Remove:    remove,
+		})
+	}
+	// STEP 2: non overlap - simply remove
+	for principal, remove := range remote {
+		_, ok := configured[principal]
+		if ok { // already handled in STEP 1
+			continue
+		}
+		diff.Changes = append(diff.Changes, permissions.UnityCatalogPermissionsChange{
+			Principal: principal,
+			Remove:    util.SetToSlice(remove),
+		})
+	}
+	// so that we can deterministic tests
+	sort.Slice(diff.Changes, func(i, j int) bool {
+		return diff.Changes[i].Principal < diff.Changes[j].Principal
+	})
+	return diff
+}
+
+// replacePermissionsForPrincipal merges removal diff of existing permissions on the platform
+func replacePermissionsForPrincipal(a permissions.UnityCatalogPermissionsAPI, securable string, name string, principal string, list permissions.UnityCatalogPermissionsList) error {
+	existing, err := a.GetPermissions(securable, name)
+	if err != nil {
+		return err
+	}
+	return a.UpdatePermissions(securable, name, diffPermissionsForPrincipal(principal, list, existing))
+}
+
+func filterPermissionsForPrincipal(in permissions.UnityCatalogPermissionsList, principal string) (out permissions.UnityCatalogPermissionsList) {
+
+	for _, v := range in.Assignments {
+		if v.Principal == principal {
+			out.Assignments = append(out.Assignments, v)
+		}
+	}
+	return
+}
+
+func ResourceGrant() *schema.Resource {
+	s := map[string]*schema.Schema{
```

Comment:
We should use `UnityCatalogPrivilegeAssignment` and `common.StructToSchema` , like in the ResourceGrants() resource

Reply:
Sure, I'll take a look. I have a question though, how would I conditionally mark these fields as `Required` & `ForcesNew` if using the reflective method ?

Reply:
Done.